File: rcs/riot-sdk/python/riot_sdk/facade/map_client.py
from __future__ import annotations

from typing import TYPE_CHECKING

from riot_sdk.core.business_response import ensure_success, require_response
from riot_sdk.core.exceptions import RiotApiException
from riot_sdk.core.maps import Map
from riot_sdk.core.stations import Station

if TYPE_CHECKING:
    from riot_sdk.facade.session import RiotSession


class MapClient:
    """Thin imap Map facade (BC-MAP-001 / ADR-sdk-0006)."""

    def __init__(self, session: RiotSession) -> None:
        self._session = session

    async def list_maps(self) -> list[Map]:
        """GET /api/imap/v1/mapInfo/getALLMapInfoExcludeMapJson (BC-MAP-001 / ADR-sdk-0005)."""
        client = self._session.create_generated_imap_client()
        response = require_response(
            await client.api.imap.v1.map_info.get_a_l_l_map_info_exclude_map_json.get(),
            "getALLMapInfoExcludeMapJson",
        )
        ensure_success(response.code, response.message)
        result = response.result or []
        return [
            Map(map_id=m.id, name=m.name)
            for m in result
            if m.id is not None and m.name
        ]

    async def list_stations(self, map_id: int) -> list[Station]:
        """GET /api/imap/v1/mapInfo/stations/{mapId} (BC-MAP-002 / ADR-sdk-0005)."""
        client = self._session.create_generated_imap_client()
        response = require_response(
            await client.api.imap.v1.map_info.stations.by_map_id(map_id).get(),
            f"stations/{map_id}",
        )
        ensure_success(response.code, response.message)
        result = response.result or []
        return [
            Station(map_id=map_id, station_id=s.id, name=s.name)
            for s in result
            if s.id is not None and s.name
        ]

    async def list_stations_strict(self, map_id: int) -> list[Station]:
        """Reject malformed or duplicate station rows instead of filtering them."""
        if map_id <= 0:
            raise ValueError("map_id must be positive")

        client = self._session.create_generated_imap_client()
        response = require_response(
            await client.api.imap.v1.map_info.stations.by_map_id(map_id).get(),
            f"stations/{map_id}",
        )
        ensure_success(response.code, response.message)
        if response.result is None:
            raise RiotApiException(
                f"stations/{map_id} returned null result.",
                business_code="station-catalog-missing",
            )
        if not response.result:
            raise RiotApiException(
                f"stations/{map_id} returned an empty station catalog.",
                business_code="station-catalog-empty",
            )

        station_ids: set[int] = set()
        stations: list[Station] = []
        for row in response.result:
            if row.id is None or row.id <= 0 or not row.name or not row.name.strip():
                raise RiotApiException(
                    f"stations/{map_id} returned a malformed station row.",
                    business_code="station-catalog-entry-invalid",
                )
            if row.id in station_ids:
                raise RiotApiException(
                    f"stations/{map_id} returned duplicate station id {row.id}.",
                    business_code="station-catalog-duplicate",
                )
            station_ids.add(row.id)
            stations.append(Station(map_id=map_id, station_id=row.id, name=row.name))
        return stations

    @property
    def raw(self):
        """Underlying Kiota imap client for endpoints not yet wrapped."""
        return self._session.create_generated_imap_client()

File: rcs/riot-sdk/csharp/RIoT.Sdk.Facade/MapClient.cs
using RIoT.Sdk.Core;

namespace RIoT.Sdk.Facade;

/// <summary>
/// Thin imap Map facade (BC-MAP-001 / ADR-sdk-0006).
/// </summary>
public sealed class MapClient
{
    private readonly RiotSession _session;

    internal MapClient(RiotSession session) => _session = session;

    /// <summary>
    /// GET /api/imap/v1/mapInfo/getALLMapInfoExcludeMapJson — list Maps without mapJson (BC-MAP-001).
    /// Throws <see cref="RiotApiException"/> on business failure (ADR-sdk-0005).
    /// </summary>
    public async Task<IReadOnlyList<Map>> ListMapsAsync(CancellationToken cancellationToken = default)
    {
        var client = _session.CreateGeneratedImapClient();
        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Imap.V1.MapInfo.GetALLMapInfoExcludeMapJson
                .GetAsync(cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            "getALLMapInfoExcludeMapJson");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);

        var result = response.Result ?? [];
        return result
            .Where(m => m.Id is not null && !string.IsNullOrWhiteSpace(m.Name))
            .Select(m => new Map(m.Id!.Value, m.Name!))
            .ToList();
    }

    /// <summary>
    /// GET /api/imap/v1/mapInfo/stations/{mapId} — list Stations on a Map (BC-MAP-002).
    /// Throws <see cref="RiotApiException"/> on business failure (ADR-sdk-0005).
    /// Empty list on success is a valid domain result (e.g. invalid mapId=0).
    /// </summary>
    public async Task<IReadOnlyList<Station>> ListStationsAsync(
        int mapId,
        CancellationToken cancellationToken = default)
    {
        var client = _session.CreateGeneratedImapClient();
        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Imap.V1.MapInfo.Stations[mapId]
                .GetAsync(cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            $"stations/{mapId}");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);

        var result = response.Result ?? [];
        return result
            .Where(s => s.Id is not null && !string.IsNullOrWhiteSpace(s.Name))
            .Select(s => new Station(mapId, s.Id!.Value, s.Name!))
            .ToList();
    }

    /// <summary>
    /// Strict station-catalog read for safety-sensitive consumers. Unlike
    /// <see cref="ListStationsAsync"/>, malformed and duplicate rows are rejected rather
    /// than silently filtered.
    /// </summary>
    public async Task<IReadOnlyList<Station>> ListStationsStrictAsync(
        int mapId,
        CancellationToken cancellationToken = default)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(mapId);

        var client = _session.CreateGeneratedImapClient();
        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Imap.V1.MapInfo.Stations[mapId]
                .GetAsync(cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            $"stations/{mapId}");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);
        if (response.Result is null)
        {
            throw new RiotApiException(
                $"stations/{mapId} returned null result.",
                businessCode: "station-catalog-missing");
        }
        if (response.Result.Count == 0)
        {
            throw new RiotApiException(
                $"stations/{mapId} returned an empty station catalog.",
                businessCode: "station-catalog-empty");
        }

        List<Station> stations = new(response.Result.Count);
        HashSet<int> stationIds = [];
        foreach (var row in response.Result)
        {
            if (row.Id is null or <= 0 || string.IsNullOrWhiteSpace(row.Name))
            {
                throw new RiotApiException(
                    $"stations/{mapId} returned a malformed station row.",
                    businessCode: "station-catalog-entry-invalid");
            }
            if (!stationIds.Add(row.Id.Value))
            {
                throw new RiotApiException(
                    $"stations/{mapId} returned duplicate station id {row.Id.Value}.",
                    businessCode: "station-catalog-duplicate");
            }

            stations.Add(new Station(mapId, row.Id.Value, row.Name));
        }

        return stations;
    }

    /// <summary>Underlying Kiota imap client for endpoints not yet wrapped.</summary>
    public RIoT.Sdk.Generated.Imap.ImapClient Raw => _session.CreateGeneratedImapClient();
}
